xee/helpers.py

```diff
@@ -0,0 +1,216 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Helper functions for constructing pixel grid parameters for Xee.
+
+These helpers produce the three required keyword arguments passed to
+``xarray.open_dataset(..., engine='ee', **grid_params)``:
+
+* ``crs`` – The target Coordinate Reference System.
+* ``crs_transform`` – A 6-tuple affine transform (origin + scale) in CRS units.
+* ``shape_2d`` – The (width, height) pixel shape of the output grid.
+
+Two primary workflows:
+
+1. :func:`extract_grid_params` – Match the *native* grid of an Earth Engine
+  Image or ImageCollection.
+2. :func:`fit_geometry` – Derive a grid that fits a user geometry using either
+  an explicit pixel scale (``grid_scale``) or an explicit pixel shape
+  (``grid_shape``).
+
+All scale values must be expressed in the units of ``grid_crs``. For
+geographic CRSs (e.g. ``EPSG:4326``) this is degrees. For projected CRSs (e.g.
+UTM) this is meters.
+"""
+import math
+
+import affine
+import ee
+from pyproj import Transformer
+import shapely
+from shapely.ops import transform
+from typing import TypedDict, Union
+
+
+TransformType = tuple[float, float, float, float, float, float]
+ShapeType = tuple[int, int]
+ScalingType = tuple[float, float]
+
+
+class PixelGridParams(TypedDict):
+  """TypedDict describing pixel grid parameters.
+
+  - ``crs``: EPSG code or WKT for output grid CRS.
+  - ``crs_transform``: 6-tuple affine transform ``(a, b, c, d, e, f)``:
+      a = pixel width (x scale)
+      b = row rotation (usually 0)
+      c = x origin (upper-left x)
+      d = column rotation (usually 0)
+      e = pixel height (y scale, negative for north-up)
+      f = y origin (upper-left y)
+  - ``shape_2d``: ``(width, height)`` pixel counts.
+  """
+  crs: str
+  crs_transform: TransformType
+  shape_2d: ShapeType
+
+
+def set_scale(
+    crs_transform: TransformType,
+    scaling: ScalingType,
+  ) -> list:
+  """Return a new CRS transform with updated scale components.
+
+  Useful for adjusting an existing transform's pixel size while retaining its
+  origin. A negative y scale preserves north-up orientation.
+
+  Args:
+    crs_transform: Existing 6-value transform tuple.
+    scaling: ``(x_scale, y_scale)`` pair. ``y_scale`` may be negative for
+      north-up images.
+
+  Returns:
+    A list of the 6 affine transform values with updated scale components.
+
+  Raises:
+    TypeError: If ``scaling`` is not a length-2 tuple.
+  """
+  if isinstance(scaling, tuple) and len(scaling) == 2:
+    x_scale, y_scale = scaling
+    crs_transform[0] = x_scale
+    crs_transform[4] = y_scale
+  else:
+    raise TypeError(f'Expected a tuple of length 2 for scaling, got {scaling}')
+  affine_transform = affine.Affine(*crs_transform)
+  return list(affine_transform)[:6]
+
+
+def fit_geometry(
+  geometry: shapely.geometry.base.BaseGeometry,
+  *,
```

Comment:
A lot of folks do not know that this starts the required keyword section, so a comment might be good.

Reply:
Done